src/qibo/transpiler/optimizer.py

```diff
@@ -66,3 +67,261 @@ def __call__(self, circuit: Circuit):
+                    tmp_gate = type(gate)(
+                        primary_qubit, prev_gate.parameters[0] + gate.parameters[0]
+                    )
+                    previous_gates[primary_qubit] = tmp_gate
+                else:
+                    if isinstance(prev_gate, self.rotated_gates):
+                        self.gates.append(prev_gate)
+                    previous_gates[primary_qubit] = gate
+            else:
+                # Flush stored rotations before adding new gate
+                for q in gate.init_args:
+                    if isinstance(previous_gates[q], self.rotated_gates):
+                        self.gates.append(previous_gates[q])
+                        previous_gates[q] = None
+
+                self.gates.append(gate)
+                for q in gate.qubits:
+                    previous_gates[q] = gate
+
+        # Append any remaining rotation gates
+        self.gates.extend(
+            g for g in previous_gates if isinstance(g, self.rotated_gates)
+        )
+
+
+class U3GateFusion(Optimizer):
+    """Merge pairs of U3 gates in the circuit"""
+
+    def __init__(self):
+        self.gates = []
+
+    def __call__(self, circuit: Circuit) -> Circuit:
+
+        if 0 == circuit.ngates or circuit.gates_of_type(gates.FusedGate):
+            return circuit
+
+        tmp_new_circuit = circuit.copy(True)
+        self.__find_gates(tmp_new_circuit)
+
+        new_circuit = circuit.__class__(**tmp_new_circuit.init_kwargs)
+
+        for gate in self.gates:
+            new_circuit.add(gate)
+
+        return new_circuit
+
+    def __find_gates(self, newCircuit: Circuit) -> Circuit:
+        """
+        Identifies pairs of U3 gates that can be merged
+        """
+
+        previous_gates = [None] * newCircuit.nqubits
+
+        for gate in newCircuit.queue:
+
+            if not gate.qubits:
+                continue
+
+            primary_qubit = gate.init_args[0]  # Extract primary qubit
+
+            if isinstance(gate, gates.U3):
+                prev_gate = previous_gates[primary_qubit]
+                if isinstance(prev_gate, gates.U3):
+                    previous_gates[primary_qubit] = self.__U3Fusion(
+                        primary_qubit, prev_gate, gate
+                    )
+                else:
+                    previous_gates[primary_qubit] = gate
+            else:
+                # Flush stored U3 before adding new gate
+                for q in gate.init_args:
+                    if isinstance(previous_gates[q], gates.U3):
+                        self.gates.append(previous_gates[q])
+                        previous_gates[q] = None
+
+                self.gates.append(gate)
+                for q in gate.qubits:
+                    previous_gates[q] = gate  # Track current gate
+
+        # Append any remaining U3 gates in one pass
+        self.gates.extend(g for g in previous_gates if isinstance(g, gates.U3))
+
+    @staticmethod
+    def __extract_u3_params(U):
```

Comment:
It is recommended that variable names should be at least 3 letters. Ideally, more descriptive.